keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,767 @@
+be preferred; otherwise IPv4 addresses will be preferred.
+
+Although these rules can end up choosing an IPv6 Primary Node IP when
+no `--node-ip` was passed in some situations, this will not actually
+result in a fully-functioning node, since kubelet will set up
+IPv4-only iptables rules if no `--node-ip` was passed.
+
+#### IPv6 Addresses in `Node.Status.Addresses`
+
+Most cloud providers do not currently include IPv6 addresses in
+`Node.Status.Addresses`. (See the "[Cloud Provider
+Survey](#cloud-provider-survey)" at the end.) To some extent this is
+just because they haven't been updated for IPv6/Dual-Stack support,
+but this is also because people are worried about confusing existing
+clients by adding IPv6 addresses to a field that historically only
+held IPv4 addresses. (Something we have also worried about in other
+places. eg, Endpoints.)
+
+Some cloud providers (notably Azure) are already returning IPv6 node
+IPs when they exist, without apparent problems. OTOH, the fact that
+there haven't been problems with clusters on Azure (a relatively new
+platform) may not mean that there wouldn't be problems with clusters
+on AWS if the AWS CloudProvider suddenly started returning IPv6
+addresses.
+
+On the other other hand, CloudProviders will not be able to return
+IPv6 addresses unless the Node actually _has_ IPv6 addresses, and on
+most platforms it probably won't unless the user explicitly chose to
+have them. So in that case, making CloudProviders suddenly start
+returning IPv6 addresses when they exist still won't cause problems,
+even in clusters where there is old software that only handles IPv4
+addresses in `Node.Status.Addresses`, since the clusters running the
+old software presumably don't have any IPv6 node addresses anyway.
+
+If we are concerned about confusing old software, there are three
+approaches we could take to fix it:
+
+  1. Add one or more new `v1.NodeAddressType` values to use for IPv6
+     addresses. eg, `InternalIPv6` and `ExternalIPv6`. One problem
+     with this is that there is code scattered across many components
+     (and, eg, network plugins) that assumes that all Nodes have
+     either an `InternalIP` address or an `ExternalIP` address, so
+     using different types for IPv6 addresses would likely break
+     single-stack IPv6 clusters.
+
+  2. Provide CloudProvider-level configuration options to say whether
+     IPv6 node addresses should be returned or not. The OpenStack and
+     vSphere providers do this currently.
+
+  3. Filter `Node.Status.Addresses` based on desired cluster
+     configuration; if the user wants IPv6 addresses to be removed,
+     they must configure kubelet to request an obligatorily
+     single-stack IPv4 cluster.
+
+     (Of course, they only actually need to do that if they're running
+     their obligatorily single-stack IPv4 cluster on
+     actually-dual-stack hosts, which they presumably aren't.)
+
+The rest of this assumes the last approach.
```

Comment:
This assumes the IPv4 and IPv6 addresses naturally pair up in some obvious way, but in all the cloud APIs I've seen, there are separate APIs for getting the IPv4 IPs and the IPv6 IPs, so it's more natural for the cloud provider to return uncorrelated IPv4 and IPv6 addresses.

And like, what if you have an interface with 2 IPv4 addresses and 1 IPv6 address?

Reply:
> This assumes the IPv4 and IPv6 addresses naturally pair up in some obvious way, but in all the cloud APIs I've seen, there are separate APIs for getting the IPv4 IPs and the IPv6 IPs, so it's more natural for the cloud provider to return uncorrelated IPv4 and IPv6 addresses.

My thinking here is that the `InternalI` and `ExternalIP` address types are really meant for internal consumption by the kubelet and kube-apiserver. So while _sometimes_ they may be separate APIs calls from the cloud provider, they should be coupled based on the desired behavior from kube-apiserver and kubelet, not the cloud provider.  Though "InternalIP" and "ExternalIP" should still be somewhat meaningful for external consumers.

Reply:
> And like, what if you have an interface with 2 IPv4 addresses and 1 IPv6 address?

I would think we would categorize them based on "InternalIP" and "ExternalIP" and then order them based on ip family priorities (ipv4,ipv6 or ipv6,ipv4).

Reply:
Hm... ok, so I was misremembering the cloud APIs; it looks like (at least in the case of AWS, GCE, and Azure), the "get node IPs" APIs are per-interface. So you _can_ match up IPv4 and IPv6 addresses that are on the same interface... but if you're going to do that, then what do you do when there are multiple IPv4 addresses on an interface?

eg, if you have `10.0.0.1`, and `2001:1234` on `eth0` and `10.0.0.2` on `eth1`, you can return:

    Addresses:
      - Type: InternalIP
        Address: 10.0.0.1
        Addresses: [ "10.0.0.1", "2001::1234" ]
      - Type: InternalIP
        Address: 10.0.0.2
        Addresses: [ "10.0.0.2" ]
      - Type: ExternalIP
        Address: 1.2.3.4
        Addresses: [ "1.2.3.4"]

But then when if `10.0.0.2` was on `eth0` too? If we return

    Addresses:
      - Type: InternalIP
        Address: 10.0.0.1
        Addresses: [ "10.0.0.1", "10.0.0.2", "2001::1234" ]
      - Type: ExternalIP
        Address: 1.2.3.4
        Addresses: [ "1.2.3.4"]

then the `10.0.0.2` address becomes invisible to legacy users. Or would we do something like:

    Addresses:
      - Type: InternalIP
        Address: 10.0.0.1
        Addresses: [ "10.0.0.1", "10.0.0.2", "2001::1234" ]
      - Type: InternalIP
        Address: 10.0.0.2
        Addresses: [ "10.0.0.1", "10.0.0.2", "2001::1234" ]
      - Type: ExternalIP
        Address: 1.2.3.4
        Addresses: [ "1.2.3.4"]

? (Either way it seems like we really want to add an "InterfaceName" field...)

Or is this not how you were thinking of grouping addresses?

Reply:
> Or is this not how you were thinking of grouping addresses?

This looks right, though I would only expect 1 ipv4 and and 1 ipv6 address at most. If the kubelet sees mulitple of each family, we would filter it using the same logic you mentioned in the KEP. 


> ? (Either way it seems like we really want to add an "InterfaceName" field...)

Not sure about this yet, but with the current way of doing things we would have the same problem, but just three node addresses of type InternalIP instead of one right? If not I would think we would filter `Addresses []string` in the same way we're expecting `Addresses []NodeAddresses` to be filtered based on the ip family configuration.

Reply:
> If the kubelet sees mulitple of each family, we would filter it using the same logic you mentioned in the KEP.

If the extra IPs are there now, then removing it is kind of an API break. People may be depending on those extra IPs being listed there.

For the IPv6-addresses-in-single-stack-IPv4 clusters case, although the KEP talks about filtering out IPv6 addresses, we're only filtering out stuff that wouldn't have been there in older releases anyway (since most cloud providers currently return IPv4 addresses only). So relative to older releases, what's happening in this KEP is that `NodeAddresses` remains unchanged in single-stack IPv4 clusters, but in dual-stack clusters it will now have IPv6 addresses mixed in as well.

Maybe I should clarify that...

Reply:
> > If the extra IPs are there now, then removing it is kind of an API break. People may be depending on those extra IPs being listed there.

So I was assuming the extra IPs in `Addresses` are only added based on the dual-stack configuration on the kubelet based on the `--node-ips` flag (or whatever we decide to add). On existing single-stack IPv4 clusters, `Addresses` is a single length string array containing the same address as `Address`.  

> For the IPv6-addresses-in-single-stack-IPv4 clusters case, although the KEP talks about filtering out IPv6 addresses, we're only filtering out stuff that wouldn't have been there in older releases anyway (since most cloud providers currently return IPv4 addresses only). So relative to older releases, what's happening in this KEP is that NodeAddresses remains unchanged in single-stack IPv4 clusters, but in dual-stack clusters it will now have IPv6 addresses mixed in as well.

But we would only add this to the new field `Addresses` though, do we still consider that breaking?

Reply:
You said "I would only expect 1 ipv4 and and 1 ipv6 address at most." But you can't expect that. That is not how `NodeAddresses` works currently; you can have arbitrarily many addresses of every combination of IP family and AddressType.

So, eg, if the cloud APIs report that the node has an `eth0` with the internal IPs `10.0.0.1, `10.0.0.2`, and `2001::1234`, and the external IP `1.2.3.4`; and an `eth1` with the internal IP `10.1.0.1` and the external IP `5.6.7.8`, then in 1.19 an average cloud provider would report this as something like:

    [
      { Type: "InternalIP", Address: "10.0.0.1" },
      { Type: "InternalIP", Address: "10.0.0.2" },
      { Type: "ExternalIP", Address: "1.2.3.4" },
      { Type: "InternalIP", Address: "10.1.0.1" },
      { Type: "ExternalIP", Address: "5.6.7.8" },
    ]

(in particular, ignoring the IPv6 address) and then, this KEP proposes that in a dual-stack cluster, we'd also add

      { Type: "InternalIP", Address: "2001::1234" },

in there, probably between `10.0.0.2` and `1.2.3.4`.

Now, in your version, one way of adding `Addresses` would be

    [
      { Type: "InternalIP", Address: "10.0.0.1", Addresses: [ "10.0.0.1", "2001::1234" ] },
      { Type: "InternalIP", Address: "10.0.0.2" },
      { Type: "ExternalIP", Address: "1.2.3.4" },
      { Type: "InternalIP", Address: "10.1.0.1" },
      { Type: "ExternalIP", Address: "5.6.7.8" },
    ]

but that's weird because `2001::1234` goes with `10.0.0.2` as much as it goes with `10.0.0.1`. So we could do:

    [
      { Type: "InternalIP", Address: "10.0.0.1", Addresses: [ "10.0.0.1", "10.0.0.2", "2001::1234" ] },
      { Type: "ExternalIP", Address: "1.2.3.4" },
      { Type: "InternalIP", Address: "10.1.0.1" },
      { Type: "ExternalIP", Address: "5.6.7.8" },
    ]

but I feel like that's an API break because now there's no "`Address: 10.0.0.2`" entry, which some old client may have been depending on

so you could say

    [
      { Type: "InternalIP", Address: "10.0.0.1", Addresses: [ "10.0.0.1", "10.0.0.2", "2001::1234" ] },
      { Type: "InternalIP", Address: "10.0.0.2", Addresses: [ "10.0.0.1", "10.0.0.2", "2001::1234" ] },
      { Type: "ExternalIP", Address: "1.2.3.4" },
      { Type: "InternalIP", Address: "10.1.0.1" },
      { Type: "ExternalIP", Address: "5.6.7.8" },
    ]

but that seems silly, so maybe the first one is better.

But then, I think if we are changing the layout of `NodeAddresses`, it would be simpler to just go back to the `InternalIPv6` and `ExternalIPv6` idea:

    [
      { Type: "InternalIP", Address: "10.0.0.1" },
      { Type: "InternalIP", Address: "10.0.0.2" },
      { Type: "InternalIPv6", Address: "2001::1234" },
      { Type: "ExternalIP", Address: "1.2.3.4" },
      { Type: "InternalIP", Address: "10.1.0.1" },
      { Type: "ExternalIP", Address: "5.6.7.8" },
    ]

Then old clients will just ignore the "InternalIPv6" address, and new clients won't. And we don't have to worry about filtering elements out of the list.

We would need some hack for the single-stack IPv6 case (where the IPv6 addresses have to appear as "InternalIP"/"ExternalIP" instead/also for compat reasons) but that's doable.

Reply:
> You said "I would only expect 1 ipv4 and and 1 ipv6 address at most." But you can't expect that. That is not how NodeAddresses works currently; you can have arbitrarily many addresses of every combination of IP family and AddressType.

Although today you _can_ set multiple addresses of the same type, I'm not sure in practice anyone wants to do that and in the cases where it has happened, it's almost always by accident. From my experience there's no value in setting multiple addresses of type `InternalIP` especially when kubelet always just reads the first one. So while we still have to support multiple addresses of every combination of IP Family and AddressType, I'm not sure we want to optimize for it. I feel like most users really just want "use these IPv4/IPv6 addresses for internal cluster traffic to/from kubelet" (type InternalIP) and "use these IPv4/IPv6 addresses for any traffic that needs to go outside" (type ExternalIP). 

> Then old clients will just ignore the "InternalIPv6" address, and new clients won't. And we don't have to worry about filtering elements out of the list.
> We would need some hack for the single-stack IPv6 case (where the IPv6 addresses have to appear as "InternalIP"/"ExternalIP" instead/also for compat reasons) but that's doable.

I think this is a reasonable approach as well.

Reply:
So I started writing out the `InternalIPv6` / `ExternalIPv6` solution, and I think it ends up being worse overall than the "filter out IPv6 addresses in single-stack IPv4 clusters" solution.

1. You have to rewrite the `InternalIPv6`/`ExternalIPv6` addresses to `InternalIP`/`ExternalIP` in single-stack IPv6 clusters, for compatibility, which ends up meaning that it's not _really_ "IPv4 IPs are InternalIP/ExternalIP and IPv6 IPs are InternalIPv6/ExternalIPv6", but rather "In some clusters IPv4 and IPv6 addresses are different types and in other clusters they're the same type" which is ugly.
2. If IPv6 addresses are a new type then old clients implementing the "determine Primary Node IP" algorithm on their own will always end up picking an IPv4 IP (since they won't see the IPv6 ones), which will be wrong in a dual-stack-IPv6-primary cluster.

So I still think "filter out IPv6 addresses if the user requested single-stack IPv4" ends up being the cleanest solution.